Thesis.org

```diff
@@ -1,11 +1,725 @@
+We choose that function, among other things, because it's ~key:value~ structure is well understood and is much easier to read than nested anonymous function calls. However, to achieve our goal we need to make use of a ClojureScript feature called macros.
+*** Why are macros used here?
+From now on, the previously defined functions like ~then~, ~return~, ~runner~ will be called primitives. While these enable us to compose sequentially, they don't work like ~let~ does. Listing 17 shows the current usage of ~then~.
+#+CAPTION: Sequential composition using primitive ~then~ with anonymous function calls.
+#+begin_src clojure
+  (then first
+	(fn [a]
+	  (then second
+		(fn [b]
+		  (then third
+			(fn [c]
+			  (done [a b c])))))))
+#+end_src
+The developer needs to create the anonymous functions by hand, everytime. Thankfully, the functionality of binding ~keys~ to ~values~ stays the same, as results of the components, like ~first~, will be bound to the parameter in the ~continuation~ (e.g the symbol ~a~).
+Because this is functionaly identical, we can write a macro to translate the code example above into the desired ~let~ like structure.
+*** What are macros?
+Macros are a powerful feature which lets us rewrite our ClojureScript code before it's being evaluated. We do this by selectively telling the language what to evaluate and what to keep for later evaluation. That allows us to use all of ClojureScripts functions to manipulate the input code. This is made possible partly because ClojureScript is a Lisp, so the code already looks like a Clojure data structure. The language uses this to its advantage to operate on itself. The return value of a macro will be a list of code, that will then be evaluated. Because of this we can use the entire language to transform our code, like we transform data, into something more usable.
+Macros can be found all over Clojure and ClojureScript. Functions like ~when~[8] and ~when-not~[9] are actually macros that rewrite themselves to a simple ~if~. Even the much discussed ~let~ function is a macro[10].
+*** Syntax
+The ability to delay evaluation of just some parts of our code is powerful. ClojureScript gives a couple of new symbols to decide how it should evaluate code. We can use these symbols inside of a ~defmacro~ function, which works similar to ~defn~ and ~def~ with which we define functions and values with.
+The following table lists the symbols:
+| Symbol | Name           | Function                          |
+| '      | Quoting        | Stop execution                    |
+| `      | Syntax quoting | Like ' but qualify with namespace |
+| ~      | Unquoting      | Start execution                   |
+If we quote something with ~\'~ or ~\`~ it will not be evaluated yet[11]. The difference between the two is that ~\`~ qualifies each expression with their full namespace. The ~\~~ symbol turns on evaluation inside of a ~\`~ again[11].
```

Comment:
Well, as I said, starting and stopping execution is kind of part of the story, but it’s not what macros are _about_. Wrapping something in a thunk (`(fn [] something)`) is also "stopping execution" and calling a thunk is also "starting execution" but thunks have nothing to do with macros.

Reply:
```suggestion
We can quote our code with ~'~ or ~`~ to tell Clojure(Script) not to evaluate it[11]. Quoting can be thought of as if we wrap our expression in literal quotes. Here is an example of quoting with JavaScript code: ~[1,2,3].sort()~ and ~"[1,2,3].sort()"~, where the last example isn't an expression anymore, but a String. In Clojure(Script) we can turn our quoted code back into an expression by using the ~~~ symbol[11].
The difference between the ~'~ and ~`~ symbol is that ~`~ qualifies each expression with their full namespace. So ~`map~ gets turned into ~cljs.core/map~ instead of just ~map~.
Syntax quoting is an important tool for writing macros as it allows us to to control how symbols are interpreted. We will see later how this is used to create macros that need to treat some symbols in a special way.
```

That thunk example explains it well. I now mention that it is about "control about how symbols are interpreted", though I don't 100% agree with that either.